VR/VRFramebuffer.cpp

```diff
@@ -0,0 +1,535 @@
+		const float angleLeft,
+		const float angleRight,
+		const float angleUp,
+		const float angleDown,
+		const float nearZ,
+		const float farZ) {
+
+	const float tanAngleLeft = tanf(angleLeft);
+	const float tanAngleRight = tanf(angleRight);
+
+	const float tanAngleDown = tanf(angleDown);
+	const float tanAngleUp = tanf(angleUp);
+
+	const float tanAngleWidth = tanAngleRight - tanAngleLeft;
+
+	// Set to tanAngleDown - tanAngleUp for a clip space with positive Y
+	// down (Vulkan). Set to tanAngleUp - tanAngleDown for a clip space with
+	// positive Y up (OpenGL / D3D / Metal).
+	const float tanAngleHeight = tanAngleUp - tanAngleDown;
+
+	// Set to nearZ for a [-1,1] Z clip space (OpenGL / OpenGL ES).
+	// Set to zero for a [0,1] Z clip space (Vulkan / D3D / Metal).
+	const float offsetZ = nearZ;
+
+	ovrMatrix4f result;
+	if (farZ <= nearZ) {
+		// place the far plane at infinity
+		result.M[0][0] = 2 / tanAngleWidth;
+		result.M[0][1] = 0;
+		result.M[0][2] = (tanAngleRight + tanAngleLeft) / tanAngleWidth;
+		result.M[0][3] = 0;
+
+		result.M[1][0] = 0;
+		result.M[1][1] = 2 / tanAngleHeight;
+		result.M[1][2] = (tanAngleUp + tanAngleDown) / tanAngleHeight;
+		result.M[1][3] = 0;
+
+		result.M[2][0] = 0;
+		result.M[2][1] = 0;
+		result.M[2][2] = -1;
+		result.M[2][3] = -(nearZ + offsetZ);
+
+		result.M[3][0] = 0;
+		result.M[3][1] = 0;
+		result.M[3][2] = -1;
+		result.M[3][3] = 0;
+	} else {
+		// normal projection
+		result.M[0][0] = 2 / tanAngleWidth;
+		result.M[0][1] = 0;
+		result.M[0][2] = (tanAngleRight + tanAngleLeft) / tanAngleWidth;
+		result.M[0][3] = 0;
+
+		result.M[1][0] = 0;
+		result.M[1][1] = 2 / tanAngleHeight;
+		result.M[1][2] = (tanAngleUp + tanAngleDown) / tanAngleHeight;
+		result.M[1][3] = 0;
+
+		result.M[2][0] = 0;
+		result.M[2][1] = 0;
+		result.M[2][2] = -(farZ + offsetZ) / (farZ - nearZ);
+		result.M[2][3] = -(farZ * (nearZ + offsetZ)) / (farZ - nearZ);
+
+		result.M[3][0] = 0;
+		result.M[3][1] = 0;
+		result.M[3][2] = -1;
+		result.M[3][3] = 0;
+	}
+	return result;
+}
+
+ovrMatrix4f ovrMatrix4f_CreateFromQuaternion(const XrQuaternionf* q) {
+	const float ww = q->w * q->w;
+	const float xx = q->x * q->x;
+	const float yy = q->y * q->y;
+	const float zz = q->z * q->z;
+
+	ovrMatrix4f out;
+	out.M[0][0] = ww + xx - yy - zz;
+	out.M[0][1] = 2 * (q->x * q->y - q->w * q->z);
+	out.M[0][2] = 2 * (q->x * q->z + q->w * q->y);
+	out.M[0][3] = 0;
+
+	out.M[1][0] = 2 * (q->x * q->y + q->w * q->z);
+	out.M[1][1] = ww - xx + yy - zz;
+	out.M[1][2] = 2 * (q->y * q->z - q->w * q->x);
+	out.M[1][3] = 0;
+
+	out.M[2][0] = 2 * (q->x * q->z - q->w * q->y);
+	out.M[2][1] = 2 * (q->y * q->z + q->w * q->x);
+	out.M[2][2] = ww - xx - yy + zz;
+	out.M[2][3] = 0;
+
+	out.M[3][0] = 0;
+	out.M[3][1] = 0;
+	out.M[3][2] = 0;
+	out.M[3][3] = 1;
+	return out;
+}
+
+
+/// Use left-multiplication to accumulate transformations.
+ovrMatrix4f ovrMatrix4f_Multiply(const ovrMatrix4f* a, const ovrMatrix4f* b) {
```

Comment:
It would be technically possible but too much effort at the moment. I am picking up these methods from Oculus OpenXR SDK sample code. It makes the work easier if I can test anything by copying peaces of the code in the sample project. Is this change required now?

Reply:
Okay, fair enough.

-[Unknown]